### PR TITLE
Add MapDataset.from_images

### PR DESCRIPTION
**Pull request by AtreyeeS**
**Description**

This pull request addresses #2913 

**Dear reviewer**

### CODE REVIEW

File: gammapy/datasets/map.py [hunk 1]
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
         dataset = cls.read(filename, name=data["name"], lazy=lazy, cache=cache)
         return dataset
 
+    @classmethod
+    def from_images(cls, images, energy_min, energy_max, name=None, **kwargs):
+        """Create a MapDataset from a set 2D images which do not have an
```
Comment (registerrier) [inline gammapy/datasets/map.py:1139]: from a dictionary of

File: gammapy/datasets/map.py [hunk 2]
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
...
+        energy axis. An appropriate energy axis is added for all the images.
+
+        This can be useful for modelling and fitting data reduced outside gammapy
+
+        Parameters
+        ----------
+        images: dict
+        Dictionary of images. Allowed keywords are `counts`, `background`,
```
Comment (registerrier) [inline gammapy/datasets/map.py:1147]: need to increase indentation here

File: gammapy/datasets/map.py [hunk 3]
```diff
@@ -1893,6 +1959,57 @@ def from_geoms(
...
+         --------
+         datasetonoff: `MapDatasetOnOff`
+            Map dataset.
+         """
+
+        dataset = super().from_images(
+            images, energy_min, energy_max, name=None, **kwargs
+        )
+
+        energy_axis = MapAxis.from_energy_edges([energy_min, energy_max], name="energy")
```
Comment (registerrier) [inline gammapy/datasets/map.py:1993]: you can retrieve it from the maps contained in `dataset`

File: gammapy/datasets/map.py [hunk 4]
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
...
+        --------
+        dataset: `MapDataset`
+            Map dataset.
+        """
+
+        name = make_name(name)
+        kwargs["name"] = name
+
+        energy_axis = MapAxis.from_energy_edges([energy_min, energy_max], name="energy")
+        energy_axis_true = MapAxis.from_energy_edges(
```
Comment (registerrier) [inline gammapy/datasets/map.py:1167]: What if the user has a different binning for psf kernel and exposure?

**Review by registerrier — COMMENTED** [review]
Thanks @AtreyeeS .

This looks good. But I have a problem with the energy axis definition. Here the only supported case is when energy and energy true axes are identical. Isn't it an issue?

**Comment by AtreyeeS** [comment]
Hi @registerrier : Actually I thought about this as well - Then it would need an energy dispersion kernel. Can that normally be available without having an energy axis on the exposure?

File: gammapy/datasets/map.py (hunk 4 above)
Comment (adonath) [reply gammapy/datasets/map.py:1167]: But we always require that the PSF kernel and exposure have the same energy axis, except if the PSF is in reco energy. Is that what you meant?

File: gammapy/datasets/map.py [hunk 5]
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
...
+    @classmethod
+    def from_images(cls, images, energy_min, energy_max, name=None, **kwargs):
+        """Create a MapDataset from a set 2D images which do not have an
+        energy axis. An appropriate energy axis is added for all the images.
+
+        This can be useful for modelling and fitting data reduced outside gammapy
+
+        Parameters
+        ----------
+        images: dict
```
Comment (adonath) [inline gammapy/datasets/map.py:1146]: Maybe the documentation and usage becomes simpler (especially for the `PSFKernel`  case) if the maps are passed directly like: `.from_images(counts, background, exposure, psf_kernel,  )`. I think there is also some documentation missing...

**Comment by AtreyeeS** [comment]
I removed the lines from the main code base and added them as an example in the H.E.S.S. notebook. More than think should be as internal documentation I guess.

**Comment by adonath** [comment]
Thanks @AtreyeeS for the update. There are a two things I'm not yet happy with:
- I think we should not advertise to use the `MapDatasetOnOff` for spatial fitting, even within HESS I think this has never been done. Instead just use `MapDataset` with the `BACKGROUNDMAP`
- The way the exposure is currently computed from `EXPGAMMAMAP` is incorrect I think. To get correct flux values for fitting  of models, I think one needs to (a) define the energy axis with `e_min = 1 TeV` and `e_max = np.inf * u.TeV` (or at  least a very large number...) and (b) always assume a power-law with spectral index of 2.3 or whatever index  has been chosen respectively.

**Comment by AtreyeeS** [comment]
So do you propose
- add a `datasetonoff.to_mapdataset()` before going for fitting, or directly creating a `MapDataset`? I would slightly prefer the former.
- Energy range can be changed, but there is no Spectral Model involved at this stage. Where do I incorporate this?

**Comment by adonath** [comment]
> add a datasetonoff.to_mapdataset() before going for fitting, or directly creating a MapDataset? I would slightly prefer the former.

I think I would directly create a `MapDataset` here, because the only thing one could do with the `MapDatasetOnOff` is to use the `ExcessMapEstimator`, however all the maps computed by the `ExcessMapEstimator` are already present in the HAP maps. So I think the only fair use case it fitting... 

> Energy range can be changed, but there is no Spectral Model involved at this stage. Where do I incorporate this?

I think this has to remain a comment / note at this stage...

File: docs/tutorials/data/hess.ipynb [hunk 6]
```diff
@@ -12,7 +12,17 @@
     "\n",
```
Comment (registerrier) [inline docs/tutorials/data/hess.ipynb:12]: <p>unnecessary</p><p> </p><p>Do we really want to put the Note written like this? After all we should have the opportunity to introduce data at the DL4 level.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3350/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (registerrier) [inline docs/tutorials/data/hess.ipynb:12]: <p>Do we need this? Would it really work like this?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3350/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (registerrier) [inline docs/tutorials/data/hess.ipynb:12]: <p>as explained</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3350/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

**Comment by adonath** [comment]
@AtreyeeS, what do we do about this PR?

**Comment by adonath** [comment]
Just discussed in the dev call: right now there are no public HESS maps we could use to document e.g. a morphology fit. So the proposal is to move the documentation to internal HESS documentation instead. If there is public data at some point we can add some public documentation to Gammapy as well...

**Pull request closed without merging**